rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,436 @@
+
+Note that current oneDNN implementation just takes ISA as a maximum
+ISA.  To simplify the programming model, isa selection will not appear
+in block level APIs. However, `dnnl::set_max_cpu_isa` will still be
+effective, and can be used by the user to control the maximum isa
+level used by block level APIs.
+
+
+### Handling of architectural state 
+
+Matrix accelerators typically require to set an architectural state
+(e.g. Intel AMX or ARM SME).
+
+A few elements about those architectural states:
+- there is a per-core setting, which is typically costly in terms of
+  cycle count.
+- this state is block size dependent, so executing two brgemm kernels
+  with different shapes require an expensive reconfiguration between
+  each.
+- when matrix accelerator is armed, the max frequency for the core is
+  altered, so performance of sequential/vector code can be impacted.
+- there are some OS syscall to do in order to enable expanded stack
+  space.
+
+
+There are two options here:
+- expose two APIs, `brgemm::set_hw_context()` and
+  `brgemm::release_hw_context()`. Those can be scheduled/hoisted as
+  the user see fit, but the setup function will have to be called
+  before the brgemm::execute call for each given brgemm object. The
+  release call can happen anytime the user see fit.
+- expose only `brgemm::release_hw_context()`, and hide context
+  setting. Those setting calls would have to be hidden under existing
+  functionalities (e.g. brgemm::execute). This slightly simplifies the
+  programming model, though it removes some flexibility on the user
+  side to hoist those setting calls (e.g. to do them only once per
+  thread for whole process lifetime).
+- don't expose those functionalities. This is very risky since this
+  can impact the performance of the overall application by lowering
+  the frequency for pieces of code that don't rely on those matrix
+  accelerators.
+
+The recommendation is to go with the first option, with explicit set
+and release functions, that the user can hoist as they see fit.
+
+With respect to OS syscall, we recommend to make it transparent to
+user, by making those upon first created brgemm object that would need
+to use the matrix acceleration engine. This will ensure this call
+happens before any call to hardware instruction requiring it.
+
+## Handling of attributes
+
+First, here is the list of attributes we plan on exposing:
+- fp_math mode, in order for user to control input conversions semantics;
+- scales/zero-point in order to support upconversion before computation.
+- post-ops, in order to allow fusion of elementwise and binary
+  operations.
+
+Internally, other attributes are available to improve performance, but
+given how they rely on implicit knowledge of brgemm implementation
+internals, it is preferable to not expose those as of now.
+
+We also have internal attributes to control prefetch hints, prefetch
+distances and non-temporal stores. However those are not yet fully
```

Comment:
will you consider to expose this in the future? These knobs give us flexibility to further tune the performance.

Reply:
So far, there is no plan to expose those since there is no support internally.
The only prefetching we do internally is to prefetch the destination buffer (matrix C) at the beginning of the accumulation loop so that it is ready to accumulate result/write into. And this can be enabled under the hood, without exposing API.

For other prefetching, we would have to assess potential speedups from those before considering exposing them.

Reply:
How about temporal/non-temporal loads/stores?